next-solver enable by default (#1896)
rust-lang/blog.rust-lang.org: content/enabling-next-solver-on-nightly/index.md

+++
path = "2026/08/TODO/stabilizing-the-next-solver-trait-solver"
title = "Types Team Update and Roadmap"
authors = ["lcnr"]
aliases = ["2024/06/26/types-team-update.html"]

[extra]
team = "The Rustc Trait System Refactor Initiative"
team_url = "https://github.com/rust-lang/trait-system-refactor-initiative/"
+++

> After nearly 4 years of active development, [the next-generation trait solver](https://github.com/rust-lang/rust-project-goals/issues/113) is close to stabilization. We have enabled it by default on nightly to surface any remaining issues and plan to stabilize it in the next months. This is the largest single change to the Rust compiler since its initial release. It completely replaces the existing type system components responsible for proving where-clauses, normalizing associated types, and much more. **Please try out the latest nightly and [open an issue](https://github.com/rust-lang/rust/issues) if you encounter any bugs or regressions.**

Reviewer: >**Please try out the latest nightly and [open an issue](https://github.com/rust-lang/rust/issues) if you encounter any bugs or regressions.**

May be worth mentioning that Polonius Alpha is also enabled on nightly, so any *borrowck* issues should be reported there.

Also, I think that it's probably better to direct people to the nightly-enabled tracking issue or zulip *first*, rather than strictly opening a new issue - particularly in case of duplicates (known or new).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1896/de10e8fdc08b5f0c5e0e38b5d8ac378e6925f223..607e59ad7b3a413ac27713773d79ebbd137caaf3)*

Author: I am pointing at issues here because that gives you

<img width="742" height="350" alt="image" src="https://github.com/user-attachments/assets/75f841d0-f141-4759-ad16-28444219bfc8" />

I agree that pointing them to the issue creation page instead of the issue listing is bad as we want people to look at duplicates/see the pinned tracking issue.

I feel torn about mentioning the polonius stabilization here. I don't know where to put this and do think that by pointing people to the issues page where they see "Polonius Alpha on nightly" is good enough :thinking: also, I don't trust people to figure out whether a lifetime issue is polonius or next solver, e.g. opaque type handling can result in weird lifetime errors. I guess I won't do it :thinking:

> This is an internal component of the compiler. The main benefits of this rework will come in the future. The removal of the old implementation will unblock features such a [Type Alias Impl Trait and Return Type Notation][RTN], adding new implicit default trait bounds (e.g. [`Move` and `Forget`]), and will enable us to fix [the remaining type system unsoundnesses][Project Zero].

Reviewer: >features such a [Type Alias Impl Trait and Return Type Notation][RTN]

It's weird that these two are lumped together.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1896/de10e8fdc08b5f0c5e0e38b5d8ac378e6925f223..607e59ad7b3a413ac27713773d79ebbd137caaf3)*

Author: how so, they are the same project goal after all :>

Author: I do quite like linking to project goals here. Do you have something better to link to. I feel like tracking issues are generally worse. Alternatively the RFC

Reviewer: Okay, seems fine. Linking to the project goal feels most appropriate.

It also fixes a bunch of unnecessary errors like in the following example: [godbolt](https://rust.godbolt.org/z/MMsEjTxeE)
```rust
trait OtherTrait {
    type Assoc<'a>;
}
impl OtherTrait for u32 {
    type Assoc<'a> = &'a u32;
}

trait Trait {}
impl<T: OtherTrait> Trait for (T, for<'a> fn(<T as OtherTrait>::Assoc<'a>)) {}

fn impls<T: Trait>() {}

fn main() {
    // The old implementation failed to prove
    // the where-bound of `impls`.
    impls::<(u32, for<'a> fn(&'a u32))>();
}
```

### Compile-time performance

We've spent a lot of time on the compile-time performance of the next-generation trait solver. There have been many cases where it performed quadratically or even exponentially slower than the old solver.

> Especially the last few weeks were mainly spent on improving performance. This work was shared by many people, with major contributions by [Nick Nethercote](https://github.com/nnethercote), [Jana Dönszelmann](https://github.com/jdonszelmann), [Rémy Rakic](https://github.com/lqd), and [mira](https://github.com/inkreasing).

Reviewer 2: I prefer just "jana" I think, without my last name

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1896/de10e8fdc08b5f0c5e0e38b5d8ac378e6925f223..ae884206edff7b7aadbd1353b68f088acf3d8ec2)*

> As part of this effort, [Rémy Rakic](https://github.com/lqd) compared the performance of both implementations for the top 20,000 crates on [crates.io](https://crates.io/crates?sort=downloads). Below you can find a visualization by [Jana Dönszelmann](https://github.com/jdonszelmann) of the performance changes over the last two months.

Reviewer 2: same

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1896/de10e8fdc08b5f0c5e0e38b5d8ac378e6925f223..ae884206edff7b7aadbd1353b68f088acf3d8ec2)*

Even so, this already fixes a huge number of issues. As an underapproximation, we currently know of [more than 200 issues on GitHub fixed by this change](https://github.com/rust-lang/rust/issues?q=is%3Aissue%20state%3Aopen%20label%3Afixed-by-next-solver). This also has a significant impact on compile times; more on that later. When developing on nightly, you may accidentally rely on behavior only supported by the new trait solver.

This is an incredibly big change which results in a non-trivial amount of breakage. Most of these changes are intended improvements to type inference or the removal of undesirable behavior. We are tracking the known issues and breakage [in a pinned GitHub issue](https://github.com/rust-lang/rust/issues/160895).

## What can I do?

Please update to the latest nightly version by using `rustup update nightly` and use it to test your existing projects and libraries.

> > ⚠️ While the next-generation trait solver has been enabled on our `main` branch, this change will only be accessible on the `nighly` channel starting tomorrow. You can already test it now by providing `-Znext-solver=globally` as a command-line argument to the most recent nightly version ⚠️

Reviewer 3: ```suggestion
> ⚠️ While the next-generation trait solver has been enabled on our `main` branch, this change will only be accessible on the `nightly` channel starting tomorrow. You can already test it now by providing `-Znext-solver=globally` as a command-line argument to the most recent nightly version ⚠️
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1896/7174399af20a8d876dac3f538c53d26a713af73b..763f3820eecbaeb609abaed01507fe3be4e43bd1)*

> > ⚠️ While the next-generation trait solver has been enabled on our `main` branch, this change will only be accessible on the `nighly` channel starting from Saturday 22nd August. You can already test it now by providing `-Znext-solver=globally` as a command-line argument to the most recent nightly version ⚠️

Reviewer 3: ```suggestion
> ⚠️ While the next-generation trait solver has been enabled on our `main` branch, this change will only be accessible on the `nightly` channel starting from Saturday 22nd August. You can already test it now by providing `-Znext-solver=globally` as a command-line argument to the most recent nightly version ⚠️
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1896/7174399af20a8d876dac3f538c53d26a713af73b..4782d80ce764eecd879e9945a8029cc69a67106b)*

+++
path = "2026/08/TODO/stabilizing-the-next-solver-trait-solver"
title = "Types Team Update and Roadmap"
authors = ["lcnr"]
aliases = ["2024/06/26/types-team-update.html"]

[extra]
team = "The Rustc Trait System Refactor Initiative"
team_url = "https://github.com/rust-lang/trait-system-refactor-initiative/"
+++

After nearly 4 years of active development, [the next-generation trait solver](https://github.com/rust-lang/rust-project-goals/issues/113) is close to stabilization. We have enabled it by default on nightly to surface any remaining issues and plan to stabilize it in the next months. This is the largest single change to the Rust compiler since its initial release. It completely replaces the existing type system components responsible for proving where-clauses, normalizing associated types, and much more. **Please try out the latest nightly and [open an issue](https://github.com/rust-lang/rust/issues) if you encounter any bugs or regressions.**

> This is an internal component of the compiler. The main benefits of this rework will come in the future. The removal of the old implementation will unblock features such a [Type Alias Impl Trait and Return Type Notation][RTN], adding new implicit default trait bounds (e.g. [`Move` and `Forget`]), and will enable us to fix [the remaining type system unsoundnesses][Project Zero].

Reviewer 4: ```suggestion
This is an internal component of the compiler. The main benefits of this rework will come in the future. The removal of the old implementation will unblock features such a [Type Alias Impl Trait and Return Type Notation][RTN], allow for adding new implicit default trait bounds (e.g. [`Move` and `Forget`]), and enable us to fix [the remaining type system unsoundnesses][Project Zero].
```

> Even so, this already fixes a huge amount of issues. As an underapproximation, we currently know of [more than 200 issues on GitHub fixed by this change](https://github.com/rust-lang/rust/issues?q=is%3Aissue%20state%3Aopen%20label%3Afixed-by-next-solver). This also has a significant impact on compile times, more on that later. When developing on nightly, you may accidentally rely on behavior only supported by the new trait solver.

Reviewer 4: ```suggestion
Even so, this already fixes a huge amount of issues. As an underapproximation, we currently know of [more than 200 issues on GitHub fixed by this change](https://github.com/rust-lang/rust/issues?q=is%3Aissue%20state%3Aopen%20label%3Afixed-by-next-solver). This also has a significant impact on compile times; more on that later. When developing on nightly, you may accidentally rely on behavior only supported by the new trait solver.
```

Author: is that necessary? I have a slight dislike towards semi-colons in text

Author: or well, don't feel confident using them xd

Author: ah, well, seems fine :grin:

This is an incredibly big change which results in a non-trivial amount of breakage. Most of these changes are intended improvements to type inference or the removal of undesirable behavior. We are tracking the known issues and breakage [in a pinned GitHub issue](https://github.com/rust-lang/rust/issues/160895).

### What can I do?

Please update to the latest nightly version by using `rustup update nightly` and use it to test your existing projects and libraries. 

> Please tell us if you encounter any breakage, compile-time performance regression, or bad diagnostics. We have not spent too much time on diagnostics for the next-generation trait solver, si we would also appreciate you using this nightly for development to encounter poor diagnostics and other bugs in our error handling.

Reviewer 4: ```suggestion
Please tell us if you encounter any breakage, compile-time performance regression, or bad diagnostics. We have not spent too much time on diagnostics for the next-generation trait solver, so we would also appreciate you using this nightly for development to find poor diagnostics and other bugs in our error handling.
```
